### PR TITLE
Add pop-up extension

### PR DESCRIPTION
**Pull request by Midhil457**
Extension to open pop ups.
The extension uses the methods given by javascript (`alert()`, `confirm()`, `prompt()`)

### Alert
Alert the user which the user can respond with `OK`

### Confirm
Ask confirmation of the user. The user can respond with `OK` or `Cancel`

### Prompt
Gives a pop-up in which the user can enter text
**⚠ Prompt only works in a browser**

### Try it out
https://games.gdevelop-app.com/game-9d2c6973-1a11-49f8-9b38-70eb7a35c5da/index.html

[Prompt.zip](https://github.com/GDevelopApp/GDevelop-extensions/files/7696654/Prompt.zip)

### CODE REVIEW

**Comment by arthuro555** [comment]
The prompt doesn't seem to handle the cancel button on your example build 🤔 

!rebase

**Comment by Midhil457** [comment]
Thanks for the review! 
Made the changes

### Changes
- Add a condition to check if the response to prompt was valid (Not empty)
- Now when the user cancels teh prompt, instead of teh value being set to `null`, it's set to false

Build:
https://games.gdevelop-app.com/game-e098d301-9ffb-41ae-963f-3ed01218c4b2/index.html

Project:
[Prompt.zip](https://github.com/GDevelopApp/GDevelop-extensions/files/7697039/Prompt.zip)

**Comment by Midhil457** [comment]
!rebase

**Comment by Midhil457** [comment]
This is ready for review btw

**Comment by Bouh** [comment]
Hello @Midhil457
I made changes to the sentences to start with an informative verb.
And I made Javascript code simpler to read.

Feel free to download this version and update it if needed.

!update [pop-up.zip](https://github.com/GDevelopApp/GDevelop-extensions/files/8070458/pop-up.zip)

**Comment by Bouh** [comment]
To summarize my changes:
- Used the group name to specify the category "web browsers" for prompt
- Start the sentence with an informative verb
- Javascript code easier to read
- In extension settings, I selected the Input category

**Review by Bouh — APPROVED** [review]

**Comment by Bouh** [comment]
@GDevelopApp/extensions-team Required a second review.

File: Extensions/PopUp.json
Comment (arthuro555) [inline Extensions/PopUp.json:211]: Remove that condition, it is redundant.

File: Extensions/PopUp.json
Comment (arthuro555) [inline Extensions/PopUp.json:97]: Like confirm, make this an expression directly. If null, return an empty string.

File: Extensions/PopUp.json
Comment (arthuro555) [inline Extensions/PopUp.json:31]: Remove this too

File: Extensions/PopUp.json
Comment (arthuro555) [inline Extensions/PopUp.json:64]: Remove this too

File: Extensions/PopUp.json
Comment (arthuro555) [inline Extensions/PopUp.json:4]: Alert prompt and confirm have inconsistent casing. Respect English, write it lowercase unless it is at the start of the sentence.

**Review by arthuro555 — CHANGES_REQUESTED** [review]
The API is unnecessariliy complexified, I added comments about that. Also, even if I am asking for their removal, note for the future that variable names are to be respected in JavaScript events too: put it as a child of structure `__ExtensionName`. It is especially a problem because the ID should not override any global variable with the same name, and without any indication it could do so either.

File: Extensions/PopUp.json
Comment (Bouh) [reply Extensions/PopUp.json:174]: This is an action that opens the confirmation box.
Not sure to understand your opinion.

Here mine:
I think it should be an action because the user wants **to do** something, he asks the game engine to opens a confirmation box, then the response is saved to be checked in another extension condition.
If the user search for open a confirmation box he'll never search for it in the condition list.

Comment (Bouh) [reply Extensions/PopUp.json:174]: To me an action do something or ask for something and a condition check if the actions or expressions are valid or not.

Opening a popup box is an action.
Check the returned value is a condition.

Never a condition should do or trigger an action.

Comment (arthuro555) [reply Extensions/PopUp.json:174]: Fair point, though please don't use the id as a global variable name, use a structure instead, as the best practices dictate. I'd also use a single scene variable without an ID, since those expressions are meant to get the result directly afterwards, otherwise, all inputted data is leaked. If a user wants to keep the data longer or outside the scene, well, variables exists for that purpose.

Comment (Bouh) [reply Extensions/PopUp.json:174]: I've rewrote the extension, but yes I agree that the variable structure should be used, I ask for it often and haven't done it here lol.
Thanks for your comment!

Comment (Midhil457) [reply Extensions/PopUp.json:174]: Sorry I was gone for a bit and couldn't work on this Thanks @bouh for the changes. They made the extension much better and @ Thanks @arthuro555 for the review! I am working on it!

I agree making the variables global is a teriible idea. Could it be made clear how to init and use a structure in js? I am having some issues 😅

> I'd also use a single scene variable without an ID, since those expressions are meant to get the result directly afterwards, otherwise, all inputted data is leaked. If a user wants to keep the data longer or outside the scene, well, variables exists for that purpose.

It's just convenient to have the response persist and accessable. For instance for a login system. The user's response might be called multiple times, and it's convenient that the user original response is always available. I wanted it work similar to the tweens where you use the ID to access the tweens.

:)

Comment (Bouh) [reply Extensions/PopUp.json:174]: Yep no worries, I started to fix the last point of the review.
Should be done for this Friday :)

File: Extensions/PopUp.json
Comment (arthuro555) [reply Extensions/PopUp.json:4]: Also, a sentence in parenthesis is not considered a new phrase, so unless it is just after a dot, the first letter in the phrase there is lowercase. There should be no space between the last word of a sentence and the dot. Actually, let me just
```suggestion
  "description": "This extension adds actions to display three kinds of pop-ups.\n\n## Alert\n - Pops-up a message that will pause the game until dismissed by the player.\n\n## Confirmation\n - Pops-up a dialogue containing a message with the option to either confirm or cancel.\n\n## Prompt\n**Note that this pop-up type is unsupported on previews and PC exports.**\n - Displays a message to the player with a text field to write an answer in. The player will also have the option to either confirm or discard their input.\n",
```

File: Extensions/PopUp.json
Comment (arthuro555) [reply Extensions/PopUp.json:174]: > It's just convenient to have the response persist and accessable. For instance for a login system. The user's response might be called multiple times, and it's convenient that the user original response is always available.

Again, variables are here for that purpose. Just like a text input object's text cannot be accessed after it was deleted, after the popup is closed, the input should be given to the user and then discarded. Wanting to have a value persisted is a valid thing, but it can be easily done by storing that value in a variable. Deleting stored values is not easily doable without reading the extension source code and messing with its internal variables, which most users will never ever think of doing.

Keeping all the values stored will cause a memory leak, which most users may not want, especially if they use it for example for inputting a password: anyone that has access to the opened game can easily with the JavaScript console access the variables and read the cleartext password as that temporary value was never cleared out of memory. Another example would be loading a custom map, if you have a huge JSON describing a custom map that you paste into prompt, then this house string will occupy computer resources for nothing since after parsing it, the user no longer needs the JSON text input.

> I wanted it work similar to the tweens where you use the ID to access the tweens.

The tweens' IDs are very different, since:

1. They are needed to interact with the different currently running tweens, for example, pausing or resuming one. The popups do not have multiple long-running processes that need differentiation, there is always only one that instantly yields a result.
2. Tweens are bound to an object's lifetime, so we know that once the object is deleted, they will be deleted too. A popup's result value will never be deleted since they are bound to the game's lifetime.

Comment (arthuro555) [reply Extensions/PopUp.json:174]: > Or just explain how to use the extension in the description?

Since the description is markdown, we can also provide code snippets in the form of screenshots to clarify the usage. I have done this with the advanced P2P event handling extension for example.

**Comment by Bouh** [comment]
All points has been adressed in the latest commit.
Here a project example to test the extension:
[Project game.zip](https://github.com/GDevelopApp/GDevelop-extensions/files/8141971/game.zip)

**Review by arthuro555 — APPROVED** [review]
I cleaned up the javascript code just a bit - fyi, don't use `variable.addChild("childName", new gdjs.Variable())` - GDevelop will automatically create non existing children when using `variable.getChild`

**Pull request merged**
